### PR TITLE
[AIRFLOW-4419] Refine concurrency check in scheduler

### PR DESCRIPTION
**Pull request by KevinYang21**
### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-4419) issues and references them in the PR title.
### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

* Optimized the pool, dag and task concurrency calculation logic to make it faster. Also fix a BUG where current task concurrency should not increase if the executor already has the task.

* Added helper method in pool.py to access default pool size and optimized open slot querying method fo r performance.

* **Behavior change:** Previously scheduler will schedule at most `non_pooled_task_slot_count` non-pooled tasks per scheduler loop. After this change, scheduler will schedule at most `non_pooled_task_slot_count` non-pooled tasks in total.
  * I think this is a more intuitive interpretation of the config according to [its comment](https://github.com/KevinYang21/incubator-airflow/blob/kevin_yang_4419/airflow/config_templates/default_airflow.cfg#L121-L122).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

* Update `tests.test_jobs:SchedulerJobTest.test_find_executable_task_instances_task_concurrency` to include the use case for the BUG.
* Added `tests.models.test_pool:PoolTest`.
* Added `tests.test_jobs:SchedulerJobTest.test_find_executable_task_instances_in_non_pool`

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: airflow/jobs.py [hunk 1]
```diff
@@ -1014,9 +1013,10 @@ def _change_state_for_tis_without_dagrun(self,
             )
 
     @provide_session
-    def __get_task_concurrency_map(self, states, session=None):
+    def __get_concurrency_maps(self, states, session=None):
         """
-        Returns a map from tasks to number in the states list given.
+        Returns a map from dags to number in the states list given and
```
Comment (milton0825) [inline airflow/jobs.py]:
```suggestion
        Returns a map from dags to number of task instances in the states list given and
```

File: tests/test_utils/decorators.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+# specific language governing permissions and limitations
+# under the License.
+import functools
+
+from mock import Mock
+
+from airflow import conf
+
+
+# To be removed, this is for Python2.7
```
Comment (milton0825) [inline tests/test_utils/decorators.py]: Now that we are on python 3 should be remove this?

**Review by milton0825 — APPROVED** [review]

File: tests/test_utils/decorators.py (hunk 2 above)
Comment (KevinYang21) [reply tests/test_utils/decorators.py]: Good call. Yup you are right I was using this class for my local 2.7 env

**Comment by feng-tao** [comment]
will take a look at the pr tomorrow

File: airflow/models/pool.py [hunk 3]
```diff
@@ -32,9 +33,21 @@ class Pool(Base):
...
 
+    @staticmethod
+    @provide_session
+    def default_pool_open_slots(session):
+        from airflow.models import TaskInstance as TI  # To avoid circular imports
+        total_slots = conf.getint('core', 'non_pooled_task_slot_count')
+        used_slots = session.query(func.count()).filter(
+            TI.pool == Pool.default_pool_name).filter(
+            TI.state.in_([State.RUNNING, State.QUEUED])).scalar()
+        return total_slots - used_slots
```
Comment (feng-tao) [inline airflow/models/pool.py:49]: Will we have any race issue here? Do we do a guard to make sure `total_slots - used_slots` is always larger or equal than 0?

**Review by feng-tao — COMMENTED** [review]
Didn't look at the test, but the code looks good to me

File: airflow/models/pool.py (hunk 3 above)
Comment (KevinYang21) [reply airflow/models/pool.py:49]: Yes we may have race issue and other cases causing the result to be negative( e.g. force triggered backfills). Right now the usage of this method does check the result using `<= 0` so even if we got negative returns we're fine. And logically I think it makes more sense to return negative values if we're overwhelming the pool. What do you think?

Comment (feng-tao) [reply airflow/models/pool.py:49]: lgtm

**Comment by ashb** [comment]
Worth pulling in to 1.10.4? Safe/backwards-compatible to do so?

**Comment by KevinYang21** [comment]
@ashb I believe it is safe.

File: airflow/models/pool.py [hunk 4]
```diff
@@ -43,42 +56,14 @@ def to_json(self):
             'description': self.description,
         }
 
-    @provide_session
-    def used_slots(self, session):
```
Comment (ashb) [inline airflow/models/pool.py:47]: This is used by the UI so I think we need to bring these back:

https://github.com/apache/airflow/blob/6952b196863aeb50632656bb4004129159ff573e/airflow/www/views.py#L2067-L2083

**Pull request merged**